src/Lykke.Service.BlockchainWallets.Services/BlockchainExtensionsService.cs

```diff
@@ -0,0 +1,177 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Common.Log;
+using JetBrains.Annotations;
+using Lykke.Common.Log;
+using Lykke.Service.BlockchainApi.Client;
+using Lykke.Service.BlockchainWallets.Contract;
+using Lykke.Service.BlockchainWallets.Core.DTOs;
+using Lykke.Service.BlockchainWallets.Core.Services;
+
+namespace Lykke.Service.BlockchainWallets.Services
+{
+    [UsedImplicitly]
+    public class BlockchainExtensionsService : IBlockchainExtensionsService
+    {
+        #region Fields
+        private readonly IBlockchainIntegrationService _blockchainIntegrationService;
+        private readonly ConcurrentDictionary<string, bool> _cacheCapabilities;
+        private readonly ConcurrentDictionary<string, AddressExtensionConstantsDto> _cacheConstants;
+        private readonly ILog _log;
+
+        // TODO: Add new capabilities here.
+        private readonly List<string> _supportedCapabilities = new List<string> { "IsPublicAddressExtensionRequired", "IsAddressMappingRequired" };
+
+        private const int _apiCallRetryDelay = 30; // In seconds
+        private readonly Dictionary<string, int> _blockchainConnectAttemptsDelays;
+        #endregion
+
+        #region Initialization
+        public BlockchainExtensionsService(
+            IBlockchainIntegrationService blockchainIntegrationService,
+            ILogFactory logFactory)
+        {
+            if (logFactory == null)
+                throw new ArgumentNullException(nameof(logFactory));
+            _log = logFactory.CreateLog(this);
+
+            _blockchainIntegrationService = blockchainIntegrationService ?? throw new ArgumentNullException(nameof(blockchainIntegrationService));
+
+            _cacheCapabilities = new ConcurrentDictionary<string, bool>();
+            _cacheConstants = new ConcurrentDictionary<string, AddressExtensionConstantsDto>();
+
+            _blockchainConnectAttemptsDelays = new Dictionary<string, int>();
+        }
+
+        public void FireInitializationAndForget()
+        {
+            var apiCLientsEnumerator = _blockchainIntegrationService.GetApiClientsEnumerator();
+
+            while (apiCLientsEnumerator.MoveNext())
+            {
+#pragma warning disable 4014
+                // TODO: we could collect the tasks and monitor their state by timer for more valuable error handling and logging
+                // when some of api calls fails in runtime.
+                InitializeOneAsync(apiCLientsEnumerator.Current.Key, apiCLientsEnumerator.Current.Value);
+#pragma warning restore 4014
+            }
+        }
+
+        private async Task InitializeOneAsync(string blockchainType, BlockchainApiClient apiClient)
+        {
```

Comment:
I suggest to use polly library

Reply:
Polly is greate. Thank you for information.
But the most close official example is the following: [link](https://github.com/App-vNext/Polly-Samples/blob/master/PollyDemos/Async/AsyncDemo04_WaitAndRetryForever.cs) - it does not offer a noticeable code amount reduction or dramatic code clarification. Not sure it will give an outcome - to redesign this part of code. At least, it works and is fully controlled for any further development :)